api/src/main/java/org/apache/iceberg/TableScan.java

```diff
@@ -49,6 +49,16 @@ public interface TableScan extends Scan<TableScan, FileScanTask, CombinedScanTas
    */
   TableScan asOfTime(long timestampMillis);
 
+  /**
+   * Create a new {@link TableScan} from this scan's configuration that will use the most recent
+   * snapshot as of the given snapshot ref.
+   *
+   * @param snapshotRef a snapshot Ref
+   * @return a new scan based on this with the given snapshot Ref
+   * @throws IllegalArgumentException if the snapshot cannot be found
+   */
+  TableScan useSnapshotRef(String snapshotRef);
```

Comment:
I think this this should be `useRef(String branchOrTagName)`. The term `SnapshotRef` is internal and I don't think it should be exposed.

Reply:
I think we need to separate the useBranch and useTag APIs. As you said, refs are internal. From a Spark user perspective we also want to only expose the branch/tag terms; imo I think the same case could be applied to the API level. Also considering branches can be combined with time travel we could do a separate API for that ; although there's an argument to be made to just combine useBranch + as Of Time.

Reply:
Yeah, I considered that as well. The problem is that the caller doesn't know whether the ref is a tag or a branch before calling the method. That's determined when we look at table metadata and we don't want to force the caller to do that.

There may be a better name than "ref" for `useRef`. That seems like the problem to me. Maybe we could simplify it to `use`? I'm not sure that's obvious enough.

@aokolnychyi, do you have any thoughts on the name here?

Reply:
@rdblue @amogh-jahagirdar I agree that we can use a common API for tag or branch like `useRef`.

We have two signatures:

`useRef(String refName)`

`useRef(String refName, Long timeStampMillis)` -> will throw exception for tag type, since we cant do time travel for tag.

Reply:
Sure. The only thing is  I think if we do useRef (or if we come up with a better name) then we would not want to have the `useRef(String refName, Long timeStampMillis)`. A user would chain it with the existing useTimestamp and then the validation would happen in the scan context. I don't think we would want the extra method because time travel would only apply for branches so having useRef in that case doesn't make sense to me.

If we have consensus on this, then I can update https://github.com/apache/iceberg/pull/5364/files with the updated approach. Then this, PR could be focused on the Spark side of the integration

Reply:
> will throw exception for tag type, since we cant do time travel for tag.

In that case I would suggest:

- useRef(String refName)
- useBranchAsOfTime(String branchName, Long timeStampMillis)

Reply:
Oh I see the alternative is just `.useRef(refName).asOfTime(timestampMillis)`. That also works, in that case +1 for `useRef(String refName)`

Reply:
Sounds like there is consensus for `useRef`.

Reply:
@amogh-jahagirdar  After these changes are incorporated in https://github.com/apache/iceberg/pull/5364/files, Only then this PR can be merged.